### DISCUSSION TITLE
Managed Virtual Threads

### DISCUSSION THREAD

**Discussion by jdgriffith** [oven-sh/bun#6214] [Ideas]
This is a feature ask for possibly pretty far in the future. 

Virtual threads in Java 21 and goroutines in Go offer an easy, abstracted way for developers to use threads under the hood without the manual management of OS threads. My understanding of virtual threads is it is a way to offload a task or work with concurrency built in and the developer does not have to consider the implementation of OS threads, such as the number of threads, which thread to add the work to, what thread pool implementation to use, and so on.

This is a feature which could be coded in Zig or Typescript, I believe. It could possibly even start as a third-party library but then later baked into the std library or runtime. 

IMO This is one of those features which would give Bun a big distinctive edge.

**Comment by fckffln** [comment]
+1, It would be great to have such an option, I saw such a proposal in [one issue](https://github.com/oven-sh/bun/issues/13091)

I tried to implement thread management through C++ via FFI, but as expected, when the main thread is blocked, this whole story does not work. Plus, JSCallback loses context (I would like to have a shared memory or something like a move operator)
